### PR TITLE
Add torch.unique_consecutive

### LINKED ISSUE #19045: [Feature Request] sorted_input for torch.unique
Is it good to add a `sorted_input` to `torch.unique`?

Current cuda implementation always sort the input tensor first. But sometimes the input tensor is guaranteed to be already sorted. Sorting is O(n*logn), which is likely to be the most time consuming part of the whole computation, and it would be nice if we allow user to skip this part.

`torch.unique` with `sorted_input=True` would resemble `thrust::unique` and only remove the consecutive duplicate elements.

It might be better to name it `consecutive` instead of `sorted_input`.

### PR DESCRIPTION
**Pull request by zasdfgbnm**
Fixes: https://github.com/pytorch/pytorch/issues/19045

Please review: @VitalyFedyunin @ngimel 

This is independent on the #18649 series. This will cause merge conflicts in #18649 series, but please merge this first, and I will resolve the merge conflicts there.

The new feature is exposed in `_unique2_temporary_will_remove_soon` and `_unique_dim2_temporary_will_remove_soon`. But not at `torch.unique` yet. I will take care of the API after #18649 series get merged completely.

Benchmark on a tensor of shape `torch.Size([15320, 2])`:

```python
print(torch.__version__)
print()
a = tensor.sort().values.to('cpu')
print('cpu, sorted_input=False:')
%timeit torch._unique2_temporary_will_remove_soon(a)
%timeit torch._unique2_temporary_will_remove_soon(a, return_inverse=True)
%timeit torch._unique2_temporary_will_remove_soon(a, return_counts=True)
%timeit torch._unique2_temporary_will_remove_soon(a, return_inverse=True, return_counts=True)
print()
print('cpu, sorted_input=True:')
%timeit torch._unique2_temporary_will_remove_soon(a, sorted_input=True)
%timeit torch._unique2_temporary_will_remove_soon(a, sorted_input=True, return_inverse=True)
%timeit torch._unique2_temporary_will_remove_soon(a, sorted_input=True, return_counts=True)
%timeit torch._unique2_temporary_will_remove_soon(a, sorted_input=True, return_inverse=True, return_counts=True)
print()
a = a.to('cuda')
print('cuda, sorted_input=False:')
%timeit torch._unique2_temporary_will_remove_soon(a); torch.cuda.synchronize()
%timeit torch._unique2_temporary_will_remove_soon(a, return_inverse=True); torch.cuda.synchronize()
%timeit torch._unique2_temporary_will_remove_soon(a, return_counts=True); torch.cuda.synchronize()
%timeit torch._unique2_temporary_will_remove_soon(a, return_inverse=True, return_counts=True); torch.cuda.synchronize()
print()
print('cuda, sorted_input=True:')
%timeit torch._unique2_temporary_will_remove_soon(a, sorted_input=True); torch.cuda.synchronize()
%timeit torch._unique2_temporary_will_remove_soon(a, sorted_input=True, return_inverse=True); torch.cuda.synchronize()
%timeit torch._unique2_temporary_will_remove_soon(a, sorted_input=True, return_counts=True); torch.cuda.synchronize()
%timeit torch._unique2_temporary_will_remove_soon(a, sorted_input=True, return_inverse=True, return_counts=True); torch.cuda.synchronize()
``` 

```
1.1.0a0+2addccc

cpu, sorted_input=False:
340 µs ± 5.88 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
717 µs ± 14.9 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
52.3 ms ± 2.75 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)
52.3 ms ± 1.79 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)

cpu, sorted_input=True:
32.8 µs ± 285 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
49.9 µs ± 557 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
51.6 µs ± 1.08 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)
78 µs ± 782 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)

cuda, sorted_input=False:
213 µs ± 1.52 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
291 µs ± 3.81 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
250 µs ± 1.05 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
321 µs ± 1.59 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)

cuda, sorted_input=True:
45.6 µs ± 2.13 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)
110 µs ± 2.47 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)
82 µs ± 857 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
143 µs ± 409 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
```

```python
print(torch.__version__)
print()
a1, a2 = tensor.unbind(1)
indices = (a1 * tensor.max() + a2).sort().indices 
a = tensor.index_select(0, indices).to('cpu')
print('cpu, sorted_input=False:')
%timeit torch._unique_dim2_temporary_will_remove_soon(a, dim=0)
%timeit torch._unique_dim2_temporary_will_remove_soon(a, dim=0, return_inverse=True)
%timeit torch._unique_dim2_temporary_will_remove_soon(a, dim=0, return_counts=True)
%timeit torch._unique_dim2_temporary_will_remove_soon(a, dim=0, return_inverse=True, return_counts=True)
print()
print('cpu, sorted_input=True:')
%timeit torch._unique_dim2_temporary_will_remove_soon(a, dim=0, sorted_input=True)
%timeit torch._unique_dim2_temporary_will_remove_soon(a, dim=0, sorted_input=True, return_inverse=True)
%timeit torch._unique_dim2_temporary_will_remove_soon(a, dim=0, sorted_input=True, return_counts=True)
%timeit torch._unique_dim2_temporary_will_remove_soon(a, dim=0, sorted_input=True, return_inverse=True, return_counts=True)
print()
a = a.to('cuda')
print('cuda, sorted_input=False:')
%timeit torch._unique_dim2_temporary_will_remove_soon(a, dim=0); torch.cuda.synchronize()
%timeit torch._unique_dim2_temporary_will_remove_soon(a, dim=0, return_inverse=True); torch.cuda.synchronize()
%timeit torch._unique_dim2_temporary_will_remove_soon(a, dim=0, return_counts=True); torch.cuda.synchronize()
%timeit torch._unique_dim2_temporary_will_remove_soon(a, dim=0, return_inverse=True, return_counts=True); torch.cuda.synchronize()
print()
print('cuda, sorted_input=True:')
%timeit torch._unique_dim2_temporary_will_remove_soon(a, dim=0, sorted_input=True); torch.cuda.synchronize()
%timeit torch._unique_dim2_temporary_will_remove_soon(a, dim=0, sorted_input=True, return_inverse=True); torch.cuda.synchronize()
%timeit torch._unique_dim2_temporary_will_remove_soon(a, dim=0, sorted_input=True, return_counts=True); torch.cuda.synchronize()
%timeit torch._unique_dim2_temporary_will_remove_soon(a, dim=0, sorted_input=True, return_inverse=True, return_counts=True); torch.cuda.synchronize()
```

```
cpu, sorted_input=False:
55.4 ms ± 1.12 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)
55.8 ms ± 616 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
55.2 ms ± 402 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
55.1 ms ± 725 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)

cpu, sorted_input=True:
54.7 ms ± 585 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
55.2 ms ± 1.23 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)
54.5 ms ± 865 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
54.9 ms ± 577 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)

cuda, sorted_input=False:
171 µs ± 783 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
220 µs ± 1.65 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
203 µs ± 2.95 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
251 µs ± 2.83 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)

cuda, sorted_input=True:
59.6 µs ± 757 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
113 µs ± 431 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
93.2 µs ± 2.13 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)
147 µs ± 2.81 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)
```
The CPU implementation of `unique_dim` is super slow, see https://github.com/pytorch/pytorch/issues/18987, but this PR will not worry about this issue.

### CODE REVIEW

**Comment by zasdfgbnm** [comment]
CI is broken...

**Comment by ngimel** [comment]
Looks good, but when it's finally merged it will require a documentation for `sorted_input` argument. Generally though I'm unsure if we need this option - it complicates API, it is error prone (user can pass sorted=True when input is in fact unsorted), and with a slight perf impact the implementation itself can verify if input is sorted or not, and not call sort if input is sorted.  
cc @jcjohnson @ezyang for the general API question. 
Also, while you are working with unique, can you please look into implementing #16330?

**Comment by zasdfgbnm** [comment]
> Looks good, but when it's finally merged it will require a documentation for sorted_input argument.

Yes, I will document this when it's ready to modify API.

> Generally though I'm unsure if we need this option - it complicates API, it is error prone (user can pass sorted=True when input is in fact unsorted), and with a slight perf impact the implementation itself can verify if input is sorted or not, and not call sort if input is sorted.

Not exactly, `sorted_input` actually has another benefit: if the user want the same behavior as `thrust::unique` or `std::unique`, then the user could set `sorted_input=True` to easily achieve this.
```
[1, 1, 2, 2, 1, 1, 2, 2, 3] --> [1, 2, 1, 2, 3]
```

The `sorted` argument is even worse: It basically tells ATen that it might use hashtable to optimize performance, but unfortunately only CPU implementation of `unique` cares about it. GPU implementation of `unique_dim`, `unique` and CPU implementation of `unique_dim` doesn't care about it


> cc @jcjohnson @ezyang for the general API question.

Yes, please comment.

> Also, while you are working with unique, can you please look into implementing #16330?

Sure, no problem.

**Comment by ngimel** [comment]
I see. I'd say that passing `sorted_input=True` for an unsorted input to achieve this result is abuse of the API, but it might have its uses.

**Comment by zasdfgbnm** [comment]
@ngimel Yes, kind of abuse in its current design. But I prefer to design the API in a way that, users clearly knows that this is not an abuse, and encourage them to use this way when needed. Not sure how to do this though. Open to suggestions :)

File: aten/src/ATen/native/Unique.cpp [hunk 1]
```diff
@@ -17,43 +17,87 @@ template <typename scalar_t>
 std::tuple<Tensor, Tensor, Tensor> _unique_cpu_template(
     const Tensor& self,
     const bool sorted,
```
Comment (VitalyFedyunin) [inline aten/src/ATen/native/Unique.cpp]: having `sorted` and `sorted_input` is really confusing. Also It is really two completely different function codes and I prefer to see them separated into subfunctions.

File: aten/src/ATen/native/Unique.cpp [hunk 2]
```diff
@@ -101,19 +146,21 @@ std::tuple<Tensor, Tensor, Tensor> _unique_dim_cpu_template(
...
+          }
         }
-      }
-      return false;
-    });
+        return false;
+      });
+  }
 
   Tensor input_sorted = at::empty(input_flat.sizes(), input_flat.options());
```
Comment (VitalyFedyunin) [inline aten/src/ATen/native/Unique.cpp]: `sorted_input` and `input_sorted` look terrible together.

File: aten/src/ATen/native/Unique.cpp [hunk 3]
```diff
@@ -101,19 +146,21 @@ std::tuple<Tensor, Tensor, Tensor> _unique_dim_cpu_template(
...
   for (int i = 0; i < indices.size(); ++i) {
```
Comment (VitalyFedyunin) [inline aten/src/ATen/native/Unique.cpp]: Maybe just clone if it is already sorted?

**Review by VitalyFedyunin — COMMENTED** [review]
What will happen if I pass sorted_input and randomly shuffled tensor? Garbage in - garbage out or sygfault?

**Comment by zasdfgbnm** [comment]
@VitalyFedyunin Actually it is not garbage in, if `sorted_input` is enabled, then it will have the same behavior as `std::unique` and `thrust::unique`
```
[1, 1, 2, 2, 1, 1, 2, 2, 3] --> [1, 2, 1, 2, 3]
```

**Comment by ezyang** [comment]
I think I am generally in favor of API renaming so that it's clear that we are doing std/thrust style unique-semantics; this is much better than `sorted_input`. If you want to merge this patch as is and do the rename in a follow up, that's acceptable me as well.

I'm kind of sympathetic to @ngimel's suggestion that we automatically detect if we can do the faster thing, but that's an enhancement we can always add at some later point in time.

**Comment by zasdfgbnm** [comment]
@ezyang Yes I agree. Do you have suggestions on how to name that new operator? Does `unique_consecutive` sounds good? Since this is a new operator, I think we could simply name it as what it should be, instead of modifying `_unique2_temporary_will_remove_soon` and `_unique_dim2_temporary_will_remove_soon`.

**Comment by zasdfgbnm** [comment]
BTW, I will be very busy in the following couple days, so I might not be able to reply to this.

**Comment by ezyang** [comment]
I like `unique_consecutive`. Let's do it!

File: aten/src/ATen/native/Unique.cpp (hunk 3 above)
Comment (zasdfgbnm) [reply aten/src/ATen/native/Unique.cpp]: Actually no need to clone at all, just assign would work

**Comment by zasdfgbnm** [comment]
@VitalyFedyunin @ngimel @ezyang I have refactored this PR to create a new operator `torch.unique_consecutive`. It has documentation. The modification to the two temporary operators has been reverted. Review comments by @VitalyFedyunin has been resolved.

Please take another look: @VitalyFedyunin :)

Also, could we land this first and then work on #18649? This way the merge conflicts should be a bit easier to resolve...

**Review by ezyang — APPROVED** [review]

**Comment by ezyang** [comment]
please 'merge this please' when tests are passing, thx

**Pull request closed without merging**